src/boot/paging_32.zig

```diff
@@ -0,0 +1,31 @@
+const zasm = @import("zasm");
+
+extern var p4_table: [4096]u8;
+
+export fn enable_paging() void {
+    // Load P4 to CR3 register. (CPU uses this to access the P4 table).
+    const raw_p4_addr = @intCast(u64, @ptrToInt(&p4_table[0]));
```

Comment:
Why is it `&p4_table[0]` and not just `&p4_table`?

Reply:
The type is automatically assumed to be u64 from the `@intCast`?

Reply:
> Why is it &p4_table[0] and not just &p4_table?

`&p4_table[0]` results in a *single-item* pointer, whereas `&p4_table` results in a pointer to the array (This would be the same as taking the address of an array (`&array` has type `*[4]int`, whereas `&array[0]` has type `*int`) or a slice in Go, which gives a pointer to a [reflect.SliceHeader](https://pkg.go.dev/reflect#SliceHeader) struct).

In Zig there is also the notion of [*many-item* pointers](https://ziglang.org/documentation/master/#Pointers). If we have a slice and use the `ptr field`, we get a *many-item* pointer which essentially means a pointer to zero or more items of the slice element type. This is essentially what all pointers in C mean :P

Note: a single-item pointer and a multi-item pointer to the first element of a slice both have the same pointer address, they just have different types.

See example below :)

```zig
const std = @import("std");

pub fn main() anyerror!void {
    var array = [4]u8{ 1, 2, 3, 4 };
    var slice = array[1..3];
    foo(slice);
    bar(array);
}

pub fn foo(slice: []u8) void {
    std.debug.print("=== [ slice ] ===\n", .{});

    std.debug.print("ptrs:\n", .{});
    std.debug.print("   {any}\n", .{&slice[0]});
    std.debug.print("   {any}\n", .{&slice});
    std.debug.print("   {any}\n", .{slice.ptr});
    std.debug.print("types:\n", .{});
    std.debug.print("   {any}\n", .{@TypeOf(&slice[0])});
    std.debug.print("   {any}\n", .{@TypeOf(&slice)});
    std.debug.print("   {any}\n", .{@TypeOf(slice.ptr)});

    std.debug.print("\n", .{});
}

pub fn bar(array: [4]u8) void {
    std.debug.print("=== [ array ] ===\n", .{});

    std.debug.print("ptrs:\n", .{});
    std.debug.print("   {any}\n", .{&array[0]});
    std.debug.print("   {any}\n", .{&array});

    std.debug.print("types:\n", .{});
    std.debug.print("   {any}\n", .{@TypeOf(&array[0])});
    std.debug.print("   {any}\n", .{@TypeOf(&array)});

    std.debug.print("\n", .{});
}
```
Output:
```
$ zig run bar.zig
=== [ slice ] ===
ptrs:
   u8@7ffc9d803c2b
   []u8@7ffc9d803c00
   u8@7ffc9d803c2b
types:
   *u8
   *const []u8
   [*]u8

=== [ array ] ===
ptrs:
   u8@7ffc9d803c2a
   [4]u8@7ffc9d803c2a
types:
   *const u8
   *const [4]u8
```